### PR TITLE
Update mkFit for CMSSW_12_2_0

### PR DESCRIPTION
**Pull request by mmasciov**
### PR description:

This PR follows #35652 and updates mkFit in view of CMSSW_12_2_0.

In detail, this PR:
- updates the mkFit track seed partition function for the definition of parallel tasks;
- updates the mkFit track seed converter to account for matched hits in track seeds;
- (via cms-sw/cmsdist#7427 and cms-data/RecoTracker-MkFit#7) updates mkFit release and configurations in order to improve the overall physics performance at track-building level for tracking iterations where mkFit is enabled (as in PR #35660).

It requires cms-sw/cmsdist#7427 and cms-data/RecoTracker-MkFit#7.
Note: these PRs were updated in response to BTV results in https://its.cern.ch/jira/browse/PDMVRELVALS-136, in order to improve the track impact parameter resolution in tracks reconstructed in pixel-triplet-seeded iterations. Effects of the update on track reconstructions efficiency and fakes are small.

### PR validation:

Details of performance after this PR have been presented at TRK POG meeting on November 1st: 
https://indico.cern.ch/event/1092090/#2-mkfit-status-report

Full MTV results, including the update of cms-sw/cmsdist#7427 and cms-data/RecoTracker-MkFit#7 on November 9th:
- TTbar with PU: https://mmasciovecchio.web.cern.ch/MkFit_hitSelectionWindows_Nov21/MTV_TTbar_PU_PR35974_updateNov9_12X/
- High-pT QCD without PU: https://mmasciovecchio.web.cern.ch/MkFit_hitSelectionWindows_Nov21/MTV_highPtQCD_noPU_PR35974_updateNov9_12X/

### CODE REVIEW

**Comment by slava77** [comment]
test parameters:
- pull_request = cms-sw/cmsdist#7427,cms-data/RecoTracker-MkFit#7

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Small differences are seen in tracks, and thus in all downstream reco quantities.
Do these look in line with what is expected based on the report at the TRK POG?

https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-11-05-1100+ca13d3/46688/validateJR/all_OldVSNew_RunDoubleEG2017Cwf136p793/all_OldVSNew_RunDoubleEG2017Cwf136p793c_recoTracks_generalTracks__reRECO_obj_eta.png




https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-11-05-1100+ca13d3/46688/validateJR/all_OldVSNew_RunDoubleEG2017Cwf136p793/all_OldVSNew_RunDoubleEG2017Cwf136p793c_log10recoTracks_generalTracks__reRECO_obj_pt.png

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc [hunk 1]
```diff
@@ -82,6 +82,109 @@ namespace {
...
+
+      if (z_dir_pos) {
+        bool in_tib = barrel_pos_check(S, maxR, tib1.m_rin, tib1.m_zmax);
+        bool in_tob = barrel_pos_check(S, maxR, tob1.m_rin, tob1.m_zmax);
+
+        if (!in_tib && !in_tob) {
+          reg = TrackerInfo::Reg_Endcap_Pos;
+        } else {
+          bool in_tid = endcap_pos_check(S, maxR, tidp_rout, tidp_rin, tidp1.m_zmin - tid_z_extra);
+          bool in_tec = endcap_pos_check(S, maxR, tecp_rout, tecp_rin, tecp1.m_zmin - tec_z_extra);
```
Comment (jpata) [inline RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc:158]:
```suggestion
          const bool in_tid = endcap_pos_check(S, maxR, tidp_rout, tidp_rin, tidp1.m_zmin - tid_z_extra);
          const bool in_tec = endcap_pos_check(S, maxR, tecp_rout, tecp_rin, tecp1.m_zmin - tec_z_extra);
```

also below

**Review by jpata — COMMENTED** [review]
minor comments inline

**Comment by mmasciov** [comment]
> Small differences are seen in tracks, and thus in all downstream reco quantities. Do these look in line with what is expected based on the report at the TRK POG?
> 
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-11-05-1100+ca13d3/46688/validateJR/all_OldVSNew_RunDoubleEG2017Cwf136p793/all_OldVSNew_RunDoubleEG2017Cwf136p793c_recoTracks_generalTracks__reRECO_obj_eta.png
> 
> 
> 
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-11-05-1100+ca13d3/46688/validateJR/all_OldVSNew_RunDoubleEG2017Cwf136p793/all_OldVSNew_RunDoubleEG2017Cwf136p793c_log10recoTracks_generalTracks__reRECO_obj_pt.png 

@jpata, yes: few/small differences are expected and look in line with expectations. 

Let us note that we have updated this PR via the related PRs cms-sw/cmsdist#7427 and cms-data/RecoTracker-MkFit#7, in response to BTV validation results highlighted in https://its.cern.ch/jira/browse/PDMVRELVALS-136. The PR description is going to be updated accordingly. This update allows to improve impact parameter resolution, with minor changes to track reconstruction efficiency and fake rate. MTV results are going to be provided in the PR description.

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc [hunk 2]
```diff
@@ -82,6 +82,109 @@ namespace {
...
+
+    auto endcap_neg_check = [](const Track &S, float maxR, float rout, float rin, float zmax) -> bool {
+      bool inside = maxR > rout ? S.zAtR(rout) < zmax : (maxR > rin && S.zAtR(maxR) < zmax);
+      return inside;
+    };
+
+    for (int i = 0; i < size; ++i) {
+      const Track &S = in_seeds[i];
+
+      HitOnTrack hot = S.getLastHitOnTrack();
```
[the remark this answers is not in the document]
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc]: Done.

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc [hunk 3]
```diff
@@ -82,6 +82,109 @@ namespace {
...
+      float eta = eoh[hot.layer].GetHit(hot.index).eta();
```
[the remark this answers is not in the document]
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc]: Done.

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc [hunk 4]
```diff
@@ -82,6 +82,109 @@ namespace {
...
+      bool inside = maxR > rout ? S.zAtR(rout) > zmin : (maxR > rin && S.zAtR(maxR) > zmin);
+      return inside;
+    };
+
+    auto endcap_neg_check = [](const Track &S, float maxR, float rout, float rin, float zmax) -> bool {
+      bool inside = maxR > rout ? S.zAtR(rout) < zmax : (maxR > rin && S.zAtR(maxR) < zmax);
+      return inside;
+    };
+
+    for (int i = 0; i < size; ++i) {
```
[the remark this answers is not in the document]
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc]: Done.

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc (hunk 1 above)
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc]: Done.

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc [hunk 5]
```diff
@@ -82,6 +82,109 @@ namespace {
...
+          if (!in_tid && !in_tec) {
+            reg = TrackerInfo::Reg_Barrel;
+          } else {
+            reg = TrackerInfo::Reg_Transition_Neg;
+          }
+        }
+      }
+
+      part.m_region[i] = reg;
+      part.m_sort_score[i] = 7.0f * (reg - 2) + eta;
```
[the remark this answers is not in the document]
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc]: Done.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
It looks like the failure was not related (dd4hep, https://github.com/cms-sw/cmssw/issues/36029), but it prevented the comparisons for being run. Looks like to be safe, it's better to rerun with the next IB.

**Comment by slava77** [comment]
> It looks like the failure was not related (dd4hep, #36029), but it prevented the comparisons for being run. Looks like to be safe, it's better to rerun with the next IB.

it looks like CMSSW_12_2_X_2021-11-09-1100 will be available for tests in under one hour
(wait for https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_12_2_X_2021-11-09-1100/slc7_amd64_gcc900/-GenuineIntel/ non-gpu folder to appear)

**Comment by tvami** [comment]
Is this supposed to be also backported to 12_1_X?

**Comment by tvami** [comment]
> > It looks like the failure was not related (dd4hep, #36029), but it prevented the comparisons for being run. Looks like to be safe, it's better to rerun with the next IB.
> 
> it looks like CMSSW_12_2_X_2021-11-09-1100 will be available for tests in under one hour (wait for https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_12_2_X_2021-11-09-1100/slc7_amd64_gcc900/-GenuineIntel/ non-gpu folder to appear)

I think the resolution of this issue is still not in the 1100 IB

**Comment by tvami** [comment]
@cmsbuild , please test with https://github.com/cms-sw/cmssw/pull/36021

**Comment by slava77** [comment]
> I think the resolution of this issue is still not in the 1100 IB

why do you think so?
the  #36021 is merged in CMSSW_12_2_X_2021-11-09-1100

**Comment by mmasciov** [comment]
> Is this supposed to be also backported to 12_1_X?

@tvami I will open a PR for backport to 12_1_X once this is signed.

**Comment by jpata** [comment]
+reconstruction
- updates mkFit with corresponding small reco changes in all downstream quantities (detailed slides are linked in the description
- should be merged concurrently with https://github.com/cms-sw/cmsdist/pull/7427, https://github.com/cms-data/RecoTracker-MkFit/pull/7

**Comment by tvami** [comment]
urgent

**Comment by mmasciov** [comment]
> > Is this supposed to be also backported to 12_1_X?
> 
> @tvami I will open a PR for backport to 12_1_X once this is signed.

This is done in PR #36075.

**Comment by qliphy** [comment]
+1

File: RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc [hunk 6]
```diff
@@ -82,6 +82,115 @@ namespace {
...
+    // TrackerInfo could hold joint limits for sub-detectors.
+    const auto &L = trk_info.m_layers;
+    const float tidp_rin = std::min(L[21].m_rin, L[22].m_rin);
+    const float tidp_rout = std::max(L[21].m_rout, L[22].m_rout);
+    const float tecp_rin = std::min(L[27].m_rin, L[28].m_rin);
+    const float tecp_rout = std::max(L[27].m_rout, L[28].m_rout);
+    const float tidn_rin = std::min(L[48].m_rin, L[49].m_rin);
+    const float tidn_rout = std::max(L[48].m_rout, L[49].m_rout);
+    const float tecn_rin = std::min(L[54].m_rin, L[55].m_rin);
+    const float tecn_rout = std::max(L[54].m_rout, L[55].m_rout);
```
Comment (fwyzard) [inline RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc:109]: During the review of the Patatrack PRs we received *a lot* of comments regarding the use of "magic" numbers in the pixel code - even though most of the time we had simply modified or copied existing code.

Could you fix throughout this (and similar files, if any) the use of `4`, `10`, `21`, `48`, `27`, `54`, ... ?

Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitIterationConfigESProducer.cc:109]: @fwyzard, thanks for your comment. This has been addressed in #36246.

**Pull request merged**